src/coreclr/src/jit/gentree.cpp

```diff
@@ -12498,6 +12498,26 @@ GenTree* Compiler::gtFoldTypeCompare(GenTree* tree)
         objOp = opOther->AsCall()->gtCallThisArg->GetNode();
     }
 
+    bool                 pIsExact   = false;
+    bool                 pIsNonNull = false;
+    CORINFO_CLASS_HANDLE objCls     = gtGetClassHandle(objOp, &pIsExact, &pIsNonNull);
+
+    // if both classes are "final" (e.g. System.String[]) we can replace the comparison
+    // with `true/false` + null check.
+    if ((objCls != NO_CLASS_HANDLE) &&
+        // double check if it's final via impIsClassExact
+        impIsClassExact(objCls) && impIsClassExact(clsHnd))
+    {
+        const bool typesAreEqual = objCls == clsHnd;
```

Comment:
You need to run this through `compareTypesForEquality` instead of directly comparing the handles, like the handle-handle case above does.

Reply:
@AndyAyersMS hm... for two `string[]` with exactly the same handles `compareTypesForEquality ` returns `May` (instead of `Must`)

Reply:
That seems odd. Neither handle should be a canonical subtype so it should just turn into a handle value comparison.

Reply:
oops. everything is ok with it, returns `Must` for my case, don't know what was that